### PR TITLE
core: rip out RegionType and rework Region class

### PR DESCRIPTION
**Pull request by alwaysintreble**
Unsure if the special handling in the spoiler class can be fully ripped out so I worked around it for now. Also unhappy with how I handled getting region entrances so maybe this could be changed to a Region method instead of a function in main?

### CODE REVIEW

**Comment by alwaysintreble** [comment]
Went ahead and added a Region class rewrite to this PR,

File: worlds/alttp/Regions.py
Comment (Berserker66) [inline worlds/alttp/Regions.py:308]: I would expect the signature of these to be (world, player, name, *, **) I think that's an order already established somewhere.

Comment (Berserker66) [reply worlds/alttp/Regions.py:308]: You can reorder them project wide withe PyCharm refactor btw, don't do that by hand.

File: BaseClasses.py
Comment (Berserker66) [inline BaseClasses.py:928]: This feels like it could be done cleaner, maybe as a property?

File: BaseClasses.py
Comment (alwaysintreble) [reply BaseClasses.py:1253]: This entire function is specific for LTTP handling as far as I can tell and out of scope for this PR. Planning to revisit it after moving each of the options it parses here.

File: BaseClasses.py
Comment (Ijwu) [inline BaseClasses.py:952]: Excuse me if I'm wrong, but shouldn't property names not have verbs in them? It's implied as a getter, right?

**Review by Ijwu — COMMENTED** [review]
Seems good. Is it possible to get a generation to run for every game something like 30-50 times to ensure things look good logically? Or some other testing mechanism, if possible.

File: BaseClasses.py
Comment (alwaysintreble) [reply BaseClasses.py]: I couldn't find this mentioned specifically in PEP8 but you are correct so I've changed it.

File: BaseClasses.py
Comment (black-sliver) [inline BaseClasses.py:906]: can we rename this to _hint_text, so it matches what we have for Location, and have it open to the getattr treatment we have in Location? I'm not sure if that would break anything.

Comment (alwaysintreble) [reply BaseClasses.py:906]: i think lttp will be fine as that's what i wrote the property for but this will absolutely break OoT

**Review by black-sliver — COMMENTED** [review]
I hope you see what I mean. The current solution feels a bit cluttered to me :S

File: worlds/archipidle/Rules.py
Comment (Berserker66) [inline worlds/archipidle/Rules.py:7]: self.world is the same as the world passed here, so this parameter is not needed. Also both are multiworld

File: worlds/archipidle/__init__.py
Comment (Berserker66) [inline worlds/archipidle/__init__.py:28]: why?

File: worlds/archipidle/Rules.py
Comment (alwaysintreble) [reply worlds/archipidle/Rules.py:7]: Oh wow this is really old code if I didn't notice that. Probably going to do a pass over this world in the morning

File: worlds/archipidle/__init__.py
Comment (black-sliver) [inline worlds/archipidle/__init__.py:60]: i don't see why you would change that (as part of this PR, without providing numbers to back up why your code is better)? building the local list first and appending in one sweep is a valid way to do it. if the world maintainer has "idle commits" lying around such changes might result in needles merge conflicts for them

**Review by black-sliver — CHANGES_REQUESTED** [review]
I *do* think we should merge this sooner than later, as it removes a lot of jank from writing worlds.
However I think all of my (and zerk's) comments should be addressed plus I don't see and update to Docs. Was the current implementation never documented, or did you forget about updating world api.md?

File: worlds/dkc3/Regions.py
Comment (black-sliver) [inline worlds/dkc3/Regions.py:929]: Removing this, but leaving the `exits=None` in the function is probably not a good idea, should this ever get used.

File: worlds/hk/__init__.py
Comment (black-sliver) [inline worlds/hk/__init__.py:598]: as opposed to here, where you also remove `exits=None`

File: worlds/archipidle/Rules.py
Comment (black-sliver) [inline worlds/archipidle/Rules.py:22]: why is this being removed? can you clarify?

File: worlds/archipidle/Rules.py
Comment (black-sliver) [inline worlds/archipidle/Rules.py:36]: is this really necessary? i mean archipidle is always beatable from the start, right?

File: worlds/archipidle/Rules.py
Comment (alwaysintreble) [reply worlds/archipidle/Rules.py:22]: Because the rule there is essentially `lambda true` but it wastes a bunch of time going through all the locations.

File: worlds/archipidle/Rules.py
Comment (alwaysintreble) [reply worlds/archipidle/Rules.py:36]: yeah that's fair will revert this.

File: worlds/dkc3/Regions.py
Comment (alwaysintreble) [reply worlds/dkc3/Regions.py:929]: the reason i didn't do it here is because every call of this function has a hardcoded `exits=None`

**Comment by black-sliver** [comment]
@Berserker66 this looks good now. The one big remaining question, besides merge order, is at what point we want to have the world API semi-stable (i.e. implement backwards compat and throw Deprecation Warnings): when we merge this, the APWorlds currently distributed will not work anymore, and when the APWorlds get updated, they won't work with current stable anymore.

File: BaseClasses.py
Comment (black-sliver) [reply BaseClasses.py:1290]: I do agree with treble on this one: getting rid of the lttp details in BaseClasses / other worlds is separate from getting rid of the handling here. Rewriting the whole function is kind of a bigger task and future change to this can be made without breaking the API once this is merged.

**Comment by Berserker66** [comment]
> @Berserker66 this looks good now. The one big remaining question, besides merge order, is at what point we want to have the world API semi-stable (i.e. implement backwards compat and throw Deprecation Warnings): when we merge this, the APWorlds currently distributed will not work anymore, and when the APWorlds get updated, they won't work with current stable anymore.

Not sure yet. I want to try and find time for the .apworld installer soonish, which could do some version checking in there - and would run unittests so it would bubble up some of these issues.

**Review by Berserker66 — APPROVED** [review]

**Comment by Berserker66** [comment]
oh yeah, might as well ping @espeon65536 as the OoT changes go a bit beyond mere fixing.

**Review by black-sliver — APPROVED** [review]
LGTM, but i did not test OoT either

**Pull request merged**
